Documentation/Async.md

```diff
@@ -0,0 +1,89 @@
+
+This is creates a dilemma when implementing dependencies that are resolved asyncroniously. You can design around the DI limitations by carefully architecting your code so that the injection happens after async process is completed. This requires careful planning, increased complexity in setup. It is also prone to errors.
+
+Alternatively you can inject a intermediary object that tracks the result of the async operation. When you need to access the dependency, you can use this intermediary object to check if async task is completed and get the resulting object. With the experimental async support, we would like to provide ways to tackle this problem in Extenject. You can find Async extensions in **Plugins/Zenject/OptionalExtras/Async** folder.
+
+### <a id="example"></a>Example
+
+Lets see how we can inject async dependencies through an intermediary object. Async extensions implements `AsyncInject<T>` as this intermediary. You can use it as follows. 
+
+
+```csharp
+public class TestInstaller : MonoInstaller<TestInstaller>
+{
+    public override void InstallBindings()
+    {
+         Container.BindAsync<IFoo>().FromMethod(async () =>
+            {
+                await Task.Delay(100);
+                return (IFoo) new Foo();
+            }).AsCached();
+    }
+}
+
+public class Bar : IInitializable, IDisposable
+{
+    readonly AsyncInject<IFoo> _asyncFoo;
+    IFoo _foo;
+    public Bar(AsyncInject<IFoo> asyncFoo)
+    {
+        _asyncFoo = asyncFoo;
+    }
+
+    public void Initialize()
+    {
+        if (!asycFoo.TryGetResult(out _foo))
+        {
+            asycFoo.Completed += OnFooReady;
+        }
+    }
+       
+    private void OnFooReady(IFoo foo)
+    {
+        _foo = foo;
+    }
+
+    public void Dispose()
+    {
+        asycFoo.Completed -= OnFooReady;
+    }
+}
+```
+
+Here we use `BindAsync<IFoo>().FromMethod()` to pass an async method delegate that waits for 100 ms and then returns a newly created `Foo` object. This method can be any other method with `async Task<T> Method()` signature. `BindAsync<T>` extension provides a separate binder for async operations. This binder is currently limited to a few `FromX()` providers. Features like Pooling and Factories are not supported at the moment.
+
+With above binding `AsyncInject<IFoo>` object is added to the container. Since scope is set to `AsCached()` the operation will run only once and `AsyncInject<IFoo>` will keep the result. It is important to note that async operation will not start before this binding is resolved. If you want async operation to start immediately after installing use `NonLazy()` option. 
```

Comment:
```suggestion
With the above `AsyncInject<IFoo>` binding, the instance is added to the container. Since the scope is set to `AsCached()` the operation will run only once and `AsyncInject<IFoo>` will keep the result. It is important to note that `async` operations wont start before this binding is getting resolved. If you want `async` operation to start immediately after installing, use `NonLazy()`. 
```